app/src/main/java/unipd/se18/ocrcamera/SearchResultActivity.java

```diff
@@ -0,0 +1,407 @@
+                results.count = 0;
+                return  results;
+            }
+
+            // if the constraint is more restricting than the last, use as full
+            // list the last one and save time
+            List<String> fullList = null;
+            if(constraint.length() - lastSize > 0) {
+                fullList = lastList;
+            } else {
+                // otherwise get a copy of the full ingredients list
+                fullList = getFullIngredientsList();
+            }
+
+            //convert the constraint to upper case as the ingredients from the db use
+            //upper case only
+            final String stringConstraint = constraint.toString().toUpperCase();
+
+            //add to the suggestions any ingredient that contains the constraint
+            ArrayList<String> filteredList = new ArrayList<>();
+            for(String ingredient : fullList) {
+                if(ingredient.contains(stringConstraint)) {
+                    filteredList.add(ingredient);
+                }
+            }
+
+            //sort suggestions (if a suggestions starts with the constraint then show
+            //it before)
+            Collections.sort(filteredList, new Comparator<String> () {
+                @Override
+                public int compare(String a, String b) {
+
+                    //if both strings starts with the constraint, then show the shorter before
+                    if(a.startsWith(stringConstraint) && b.startsWith(stringConstraint))
+                        return a.length() - b.length();
+
+                    //if one of the strings starts with the constraint, then show it before
+                    //using a big value
+                    if(a.startsWith(stringConstraint))
+                        return -100;
+                    if(b.startsWith(stringConstraint))
+                        return +100;
+
+                    String[] subIngredientsA = a.split(" ");
+                    String[] subIngredientsB = b.split(" ");
+
+                    //if any of the substrings starts with the constraint, show the full string
+                    //before
+                    for(int i = 1; i < subIngredientsA.length; i++)
+                        if(subIngredientsA[i].startsWith(stringConstraint))
+                            return -50;
+                    for(int i = 1; i < subIngredientsB.length; i++)
+                        if(subIngredientsB[i].startsWith(stringConstraint))
+                            return 50;
+
+                    return a.compareToIgnoreCase(b);
+                }
+            });
+
+            results.values = filteredList;
+            results.count = filteredList.size();
+
+            //update variables for any other constraint
+            lastList = (ArrayList<String>) results.values;
+            lastSize = constraint.length();
+
+            return results;
+        }
+
+        @Override
```

Comment:
the specifications are missing

Reply:
This is an Android's method, you can find the doc here:
https://developer.android.com/reference/android/widget/Filter